r/src/altrep.cpp

```diff
@@ -777,20 +822,39 @@ struct AltrepVectorString : public AltrepVectorBase<AltrepVectorString<Type>> {
     std::shared_ptr<Array> array_;
     const StringArrayType* string_array_;
     std::string stripped_string_;
-    const bool strip_out_nuls_;
+    bool strip_out_nuls_;
     bool nul_was_stripped_;
     std::string_view view_;
   };
 
-  // Get a single string, as a CHARSXP SEXP from data2.
-  // Materialize if not done so yet, given that it is
-  // likely that there will be another call from R if there is a call (e.g. unique()),
-  // and getting a string from Array is much more costly than from data2.
+  // Get a single string as a CHARSXP SEXP
+  // Note that a previous version of this function used BEGIN_CPP11
+  // and END_CPP11, which stack-allocates 8kb for the error message.
+  // This is fine for top-level calls, but here
   static SEXP Elt(SEXP alt, R_xlen_t i) {
-    if (!Base::IsMaterialized(alt)) {
-      Materialize(alt);
+    if (Base::IsMaterialized(alt)) {
+      return STRING_ELT(Representation(alt), i);
     }
-    return STRING_ELT(Representation(alt), i);
+
+    auto altrep_data = external_pointer<ArrowAltrepData>(R_altrep_data1(alt));
+    auto resolve = altrep_data->locate(i);
+    auto array = altrep_data->chunked_array->chunk(resolve.chunk_index);
+    auto j = resolve.index_in_chunk;
+
+    SEXP s = NA_STRING;
+    RStringViewer& r_string_viewer = string_viewer();
+    r_string_viewer.SetArray(array);
+    // Note: we don't check GetBoolOption("arrow.skip_nul", false) here
+    // because it is too expensive to do so. We do set this value whenever
+    // an altrep string; however, there is a chance that this value could
+    // be out of date by the time a value in the vector is accessed.
+    r_string_viewer.reset_null_was_stripped();
+    s = r_string_viewer.Convert(j);
+    if (r_string_viewer.nul_was_stripped()) {
+      Rf_warning("Stripping '\\0' (nul) from character vector");
```

Comment:
one thing to consider about warnings is that they may be converted to errors (with `options(warn = 2)`) and so this (and maybe other uses) may end up jumping, possibly bypassing the destructor for `altrep_data`. 

Perhaps this could be done in a separate function: 

```cpp
     auto altrep_data = external_pointer<ArrowAltrepData>(R_altrep_data1(alt));
     auto resolve = altrep_data->locate(i);
     auto array = altrep_data->chunked_array->chunk(resolve.chunk_index);
     auto j = resolve.index_in_chunk;

     SEXP s = NA_STRING;
     RStringViewer& r_string_viewer = string_viewer();
     r_string_viewer.SetArray(array);
```

Reply:
Or maybe `auto altrep_data` can be replaced by something that actually grabs the `ArrowAltrepData*`, e.g. 

```cpp
ArrowAltrepData* altrep_data = external_pointer<ArrowAltrepData>(R_altrep_data1(alt)).get();
```

Then it's a pointer for sure, not a smart pointer that has a destructor